handler/oauth2/introspector_jwt.go

```diff
@@ -23,38 +23,74 @@ package oauth2
 
 import (
 	"context"
+	"time"
 
-	"github.com/pkg/errors"
+	jwtx "github.com/dgrijalva/jwt-go"
 
 	"github.com/ory/fosite"
+	"github.com/ory/fosite/token/jwt"
 )
 
-type JWTAccessTokenStrategy interface {
-	AccessTokenStrategy
-	JWTStrategy
-}
-
 type StatelessJWTValidator struct {
-	JWTAccessTokenStrategy
+	jwt.JWTStrategy
 	ScopeStrategy fosite.ScopeStrategy
 }
 
 func (v *StatelessJWTValidator) IntrospectToken(ctx context.Context, token string, tokenUse fosite.TokenUse, accessRequest fosite.AccessRequester, scopes []string) (fosite.TokenUse, error) {
-	or, err := v.JWTAccessTokenStrategy.ValidateJWT(ctx, fosite.AccessToken, token)
+	t, err := validate(ctx, v.JWTStrategy, token)
 	if err != nil {
 		return "", err
 	}
 
-	for _, scope := range scopes {
-		if scope == "" {
-			continue
-		}
+	// TODO: From here we assume it is an access token, but how do we know it is really and that is not an ID token?
```

Comment:
Hm yeah good question. I think we generally assume that ID and Access Tokens use different private keys but that check is obviously implicit. I mean we could add the `token_type` to the JWT and set it to `Bearer` (ID Tokens are not HTTP Authorization Bearer Tokens) to check this?

Reply:
> I think we generally assume that ID and Access Tokens use different private keys but that check is obviously implicit.

Ohh. Is this documented anywhere? Because this is the first time I hear this. I was passing same private key to both `NewOAuth2JWTECDSAStrategy` and `NewOpenIDConnectECDSAStrategy`. It [does not help that compose does it as well](https://github.com/ory/fosite/blob/master/compose/compose.go#L101-L104).

> I mean we could add the token_type to the JWT and set it to Bearer

Is this claim standardized anywhere? I can see it some implementations do it, but I could not find a spec?

I see that currently ID tokens have `auth_time` and access tokens do not have it. And access tokens have `scope` and ID tokens do not have it (not sure if any spec controls this). I also have `client_id` in access tokens in my implementation. In my implementation `scope` is always required but I think fosite does not enforce that.

Maybe we could just make it that `scope` claim is always present in access tokens (even if empty string) and never in ID tokens and use that?

Reply:
Personally, I would suggest that:
* Fosite introspection endpoint supports both ID tokens and access tokens (https://github.com/ory/fosite/issues/410).
  * Then here we do not have to care which one it is, we just return all claims.
* Caller of `introspect` has to check `aud` claim to verify that the token contains expected audience (which is in fact the main way to differentiate between ID tokens and access tokens, because they have different audience).
  * To make this easier, we add `audience` parameter to introspection endpoint, for caller to be able to specify expected audiences. Similar to how we currently have `scope` parameter already (both would then be a fosite extension, I have not found any spec for `scope` parameter).

Reply:
I actually did not know ID and access tokens are assumed to use different private keys. I believe we have been using the same key for ID and access tokens on our end. I think I need to do some reading up on this. Does anyone know if this is this mentioned somewhere in the OIDC RFC?

Reply:
They are fundamentally different tokens (proof of authentication, proof of authorization). They should get different keys to prevent mixing them together.

Reply:
But how you communicate to consumers that they are using different tokens? Because there is only one `/keys` endpoint where you have to publish both keys? So how consumers know which key is to be used for which tokens?

Reply:
The access token keys should not exposed for external consumption IMO.

Reply:
For JWTs of course they are? That is the whole point that you can validate JWT based access tokens without having to contact introspection endpoint.

Reply:
Introspection is internal facing, not external.

Reply:
>    This specification defines a method for a protected resource to query
   an OAuth 2.0 authorization server to determine the active state of an
   OAuth 2.0 token and to determine meta-information about this token.
   OAuth 2.0 deployments can use this method to convey information about
   the authorization context of the token from the authorization server
   to the protected resource.

Reply:
I mean, it is a protected resource. But that does not mean it has to be internal facing? It just has to be protected to prevent scanning attacks.